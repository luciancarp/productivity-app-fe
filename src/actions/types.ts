// Alert
export const ADD_ALERT = 'ADD_ALERT'
export const REMOVE_ALERT = 'REMOVE_ALERT'

export enum AlertVariants {
  success = 'success',
  error = 'error',
}

export type AlertType = {
  id: string
  variant: AlertVariants
  message: string
}

export type AlertActionTypes =
  | { type: 'ADD_ALERT'; payload: AlertType }
  | { type: 'REMOVE_ALERT'; payload: string }

// User

export const GET_USER_SUCCESS = 'GET_USER_SUCCESS'
export const GET_USER_LOADING = 'GET_USER_LOADING'
export const GET_USER_FAIL = 'GET_USER_FAIL'

export const LOGIN_USER_SUCCESS = 'LOGIN_USER_SUCCESS'
export const LOGIN_USER_LOADING = 'LOGIN_USER_LOADING'
export const LOGIN_USER_FAIL = 'LOGIN_USER_FAIL'

export const LOGOUT_USER = 'LOGOUT_USER'

export const CREATE_USER_SUCCESS = 'CREATE_USER_SUCCESS'
export const CREATE_USER_LOADING = 'CREATE_USER_LOADING'
export const CREATE_USER_FAIL = 'CREATE_USER_FAIL'

export type UserType = {
  id: string
  name: string
  email: string
  date: Date
}

export type UserActionTypes =
  | { type: 'GET_USER_SUCCESS'; payload: UserType }
  | { type: 'GET_USER_LOADING' }
  | { type: 'GET_USER_FAIL' }
  | { type: 'LOGIN_USER_SUCCESS'; payload: UserType }
  | { type: 'LOGIN_USER_LOADING' }
  | { type: 'LOGIN_USER_FAIL' }
  | { type: 'LOGOUT_USER' }
  | { type: 'CREATE_USER_SUCCESS'; payload: UserType }
  | { type: 'CREATE_USER_LOADING' }
  | { type: 'CREATE_USER_FAIL' }

// Project

export const GET_PROJECTS_SUCCESS = 'GET_PROJECTS_SUCCESS'
export const GET_PROJECTS_LOADING = 'GET_PROJECTS_LOADING'
export const GET_PROJECTS_FAIL = 'GET_PROJECTS_FAIL'

export const CREATE_PROJECT_SUCCESS = 'CREATE_PROJECT_SUCCESS'
export const CREATE_PROJECT_LOADING = 'CREATE_PROJECT_LOADING'
export const CREATE_PROJECT_FAIL = 'CREATE_PROJECT_FAIL'

export const SELECT_PROJECT = 'SELECT_PROJECT'

export const GET_PROJECT_SUCCESS = 'GET_PROJECT_SUCCESS'
export const GET_PROJECT_LOADING = 'GET_PROJECT_LOADING'
export const GET_PROJECT_FAIL = 'GET_PROJECT_FAIL'

export const DELETE_PROJECT_SUCCESS = 'DELETE_PROJECT_SUCCESS'
export const DELETE_PROJECT_LOADING = 'DELETE_PROJECT_LOADING'
export const DELETE_PROJECT_FAIL = 'DELETE_PROJECT_FAIL'

export type ProjectType = {
  id: string
  title: string
  date: Date
}

export type ProjectsType = ProjectType[]

export type ProjectActionTypes =
  | { type: 'GET_PROJECTS_SUCCESS'; payload: ProjectsType }
  | { type: 'GET_PROJECTS_LOADING' }
  | { type: 'GET_PROJECTS_FAIL' }
  | { type: 'CREATE_PROJECT_SUCCESS'; payload: ProjectType }
  | { type: 'CREATE_PROJECT_LOADING' }
  | { type: 'CREATE_PROJECT_FAIL' }
  | { type: 'SELECT_PROJECT'; payload: string }
  | { type: 'GET_PROJECT_SUCCESS'; payload: ProjectType }
  | { type: 'GET_PROJECT_LOADING' }
  | { type: 'GET_PROJECT_FAIL' }
  | { type: 'DELETE_PROJECT_SUCCESS'; payload: string }
  | { type: 'DELETE_PROJECT_LOADING' }
  | { type: 'DELETE_PROJECT_FAIL' }

// Task

export const GET_TASKS_SUCCESS = 'GET_TASKS_SUCCESS'
export const GET_TASKS_LOADING = 'GET_TASKS_LOADING'
export const GET_TASKS_FAIL = 'GET_TASKS_FAIL'

export const CREATE_TASK_SUCCESS = 'CREATE_TASK_SUCCESS'
export const CREATE_TASK_LOADING = 'CREATE_TASK_LOADING'
export const CREATE_TASK_FAIL = 'CREATE_TASK_FAIL'

export const DELETE_TASK_SUCCESS = 'DELETE_TASK_SUCCESS'
export const DELETE_TASK_LOADING = 'DELETE_TASK_LOADING'
export const DELETE_TASK_FAIL = 'DELETE_TASK_FAIL'

export const CLEAR_TASKS = 'CLEAR_TASKS'

export type TaskType = {
  id: string
  title: string
  project: string
  time: string
  done: boolean
  date: Date
}

export type TasksType = TaskType[]

export type TaskActionTypes =
  | { type: 'GET_TASKS_SUCCESS'; payload: TasksType }
  | { type: 'GET_TASKS_LOADING' }
  | { type: 'GET_TASKS_FAIL' }
  | { type: 'CREATE_TASK_SUCCESS'; payload: TaskType }
  | { type: 'CREATE_TASK_LOADING' }
  | { type: 'CREATE_TASK_FAIL' }
  | { type: 'DELETE_TASK_SUCCESS'; payload: string }
  | { type: 'DELETE_TASK_LOADING' }
  | { type: 'DELETE_TASK_FAIL' }
  | { type: 'CLEAR_TASKS' }
