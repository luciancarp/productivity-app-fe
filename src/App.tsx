import { BrowserRouter as Router, Route, Switch } from 'react-router-dom'

import { ThemeProvider } from './style/Theme'
import GlobalStyle from './style/GlobalStyle'

import Projects from './components/Projects'
import Layout from './components/Layout'

const App = () => {
  return (
    <ThemeProvider>
      <GlobalStyle />
      <Router>
        <Layout>
          <Switch>
            <Route exact path='/' component={Projects} />
          </Switch>
        </Layout>
      </Router>
    </ThemeProvider>
  )
}

export default App
