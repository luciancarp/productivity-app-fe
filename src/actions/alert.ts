import { v4 as uuid } from 'uuid'
import {
  ADD_ALERT,
  REMOVE_ALERT,
  AlertVariants,
  AlertActionTypes,
} from './types'

export function addAlert(
  variant: AlertVariants,
  message: string
): AlertActionTypes {
  return {
    type: ADD_ALERT,
    payload: {
      id: uuid(),
      variant,
      message,
    },
  }
}

export function removeAlert(id: string): AlertActionTypes {
  return {
    type: REMOVE_ALERT,
    payload: id,
  }
}
