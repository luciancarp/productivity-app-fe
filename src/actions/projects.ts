import { Dispatch } from 'redux'
import {
  PROJECTS_SUCCESS,
  PROJECTS_LOADING,
  PROJECTS_FAIL,
  ProjectsActionTypes,
  ProjectsType,
} from './types'
import axios from 'axios'

export const getProjects = () => async (
  dispatch: Dispatch<ProjectsActionTypes>
) => {
  try {
    dispatch({
      type: PROJECTS_LOADING,
    })

    const res = await axios.get(``)

    const projects: ProjectsType = res.data.data.children.map(
      (project: any) => ({
        title: project.data.title,
        date: new Date(project.data.date),
      })
    )

    dispatch({
      type: PROJECTS_SUCCESS,
      payload: projects,
    })
  } catch (e) {
    dispatch({
      type: PROJECTS_FAIL,
    })
  }
}
