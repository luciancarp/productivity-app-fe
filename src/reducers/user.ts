import {
  GET_USER_SUCCESS,
  GET_USER_LOADING,
  GET_USER_FAIL,
  LOGIN_USER_SUCCESS,
  LOGIN_USER_LOADING,
  LOGIN_USER_FAIL,
  LOGOUT_USER,
  CREATE_USER_SUCCESS,
  CREATE_USER_LOADING,
  CREATE_USER_FAIL,
  UserActionTypes,
  UserType,
} from '../actions/types'
import tokenUtility from '../utils/tokenUtility'

type InitialStateType = {
  loading: boolean
  isAuthenticated: boolean
  user?: UserType
}

const initialState: InitialStateType = {
  loading: false,
  isAuthenticated: false,
  user: undefined,
}

const userReducer = (
  state: InitialStateType = initialState,
  action: UserActionTypes
): InitialStateType => {
  switch (action.type) {
    case GET_USER_FAIL:
      return {
        ...state,
        loading: false,
      }
    case GET_USER_LOADING:
      return {
        ...state,
        loading: true,
      }
    case GET_USER_SUCCESS:
      return {
        ...state,
        loading: false,
        isAuthenticated: true,
        user: action.payload,
      }
    case LOGIN_USER_FAIL:
      return {
        ...state,
        isAuthenticated: false,
        loading: false,
      }
    case LOGIN_USER_LOADING:
      return {
        ...state,
        loading: true,
      }
    case LOGIN_USER_SUCCESS:
      return {
        ...state,
        loading: false,
        isAuthenticated: true,
        user: action.payload,
      }
    case LOGOUT_USER:
      localStorage.removeItem('token')
      tokenUtility.setAuthToken()
      return {
        ...state,
        ...initialState,
      }
    case CREATE_USER_FAIL:
      return {
        ...state,
        isAuthenticated: false,
        loading: false,
      }
    case CREATE_USER_LOADING:
      return {
        ...state,
        loading: true,
      }
    case CREATE_USER_SUCCESS:
      return {
        ...state,
        loading: false,
        isAuthenticated: true,
        user: action.payload,
      }
    default:
      return state
  }
}

export default userReducer
