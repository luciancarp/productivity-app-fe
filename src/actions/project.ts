import { Dispatch } from 'redux'
import {
  GET_PROJECTS_SUCCESS,
  GET_PROJECTS_LOADING,
  GET_PROJECTS_FAIL,
  ProjectActionTypes,
  ProjectsType,
  AlertActionTypes,
  AlertVariants,
  CREATE_PROJECT_LOADING,
  CREATE_PROJECT_SUCCESS,
  CREATE_PROJECT_FAIL,
  SELECT_PROJECT,
  GET_PROJECT_SUCCESS,
  GET_PROJECT_LOADING,
  GET_PROJECT_FAIL,
  DELETE_PROJECT_SUCCESS,
  DELETE_PROJECT_LOADING,
  DELETE_PROJECT_FAIL,
  ProjectType,
  TaskActionTypes,
  CLEAR_TASKS,
} from './types'
import axios from 'axios'
import { addAlert } from './alert'

export const getProjects =
  () => async (dispatch: Dispatch<ProjectActionTypes>) => {
    try {
      dispatch({
        type: GET_PROJECTS_LOADING,
      })

      const res = await axios.get('/api/project/user')

      const projects: ProjectsType = res.data.map((project: any) => ({
        id: project._id,
        title: project.title,
        date: new Date(project.date),
      }))

      dispatch({
        type: GET_PROJECTS_SUCCESS,
        payload: projects,
      })
    } catch (e) {
      dispatch({
        type: GET_PROJECTS_FAIL,
      })
    }
  }

export const createProject =
  (newProjectData: { title: string }) =>
  async (
    dispatch: Dispatch<ProjectActionTypes> & Dispatch<AlertActionTypes>
  ) => {
    try {
      dispatch({
        type: CREATE_PROJECT_LOADING,
      })

      const config = {
        headers: {
          'Content-Type': 'application/json',
        },
      }

      const body = JSON.stringify({
        title: newProjectData.title,
      })

      let res = await axios.post('/api/project/', body, config)

      const newProject: ProjectType = {
        id: res.data._id,
        title: res.data.title,
        date: new Date(res.data.date),
      }

      dispatch({
        type: CREATE_PROJECT_SUCCESS,
        payload: newProject,
      })
    } catch (error) {
      const errors = error.response.data.errors

      if (errors) {
        errors.forEach((error: any) => {
          dispatch(addAlert(AlertVariants.error, error.msg))
        })
      }

      dispatch({
        type: CREATE_PROJECT_FAIL,
      })
    }
  }

export const getProject =
  (id: string) => async (dispatch: Dispatch<ProjectActionTypes>) => {
    try {
      dispatch({
        type: GET_PROJECT_LOADING,
      })

      const res = await axios.get(`/api/project/${id}`)

      const project: ProjectType = {
        id: res.data._id,
        title: res.data.title,
        date: new Date(res.data.date),
      }

      dispatch({
        type: GET_PROJECT_SUCCESS,
        payload: project,
      })
    } catch (e) {
      dispatch({
        type: GET_PROJECT_FAIL,
      })
    }
  }

export const deleteProject =
  (id: string) => async (dispatch: Dispatch<ProjectActionTypes>) => {
    try {
      dispatch({
        type: DELETE_PROJECT_LOADING,
      })

      const res = await axios.delete(`/api/project/${id}`)

      const project: ProjectType = {
        id: res.data._id,
        title: res.data.title,
        date: new Date(res.data.date),
      }

      dispatch({
        type: DELETE_PROJECT_SUCCESS,
        payload: project.id,
      })
    } catch (e) {
      dispatch({
        type: DELETE_PROJECT_FAIL,
      })
    }
  }

export const selectProject =
  (id: string) =>
  (dispatch: Dispatch<ProjectActionTypes> & Dispatch<TaskActionTypes>) => {
    dispatch({
      type: SELECT_PROJECT,
      payload: id,
    })
    dispatch({
      type: CLEAR_TASKS,
    })
  }
