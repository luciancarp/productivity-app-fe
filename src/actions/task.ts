import { Dispatch } from 'redux'
import {
  GET_TASKS_SUCCESS,
  GET_TASKS_LOADING,
  GET_TASKS_FAIL,
  CREATE_TASK_SUCCESS,
  CREATE_TASK_LOADING,
  CREATE_TASK_FAIL,
  DELETE_TASK_SUCCESS,
  DELETE_TASK_LOADING,
  DELETE_TASK_FAIL,
  TaskActionTypes,
  TaskType,
  TasksType,
  AlertActionTypes,
  AlertVariants,
} from './types'
import axios from 'axios'
import { addAlert } from './alert'

export const getTasks =
  (projectId: string) => async (dispatch: Dispatch<TaskActionTypes>) => {
    try {
      dispatch({
        type: GET_TASKS_LOADING,
      })

      const res = await axios.get(`/api/task/project/${projectId}`)

      const tasks: TasksType = res.data.map((task: any) => ({
        id: task._id,
        title: task.title,
        project: task.project._id,
        time: task.time,
        done: task.done,
        date: new Date(task.date),
      }))

      dispatch({
        type: GET_TASKS_SUCCESS,
        payload: tasks,
      })
    } catch (e) {
      dispatch({
        type: GET_TASKS_FAIL,
      })
    }
  }

export const createTask =
  (newTaskData: { title: string; project: string; time: string }) =>
  async (dispatch: Dispatch<TaskActionTypes> & Dispatch<AlertActionTypes>) => {
    try {
      dispatch({
        type: CREATE_TASK_LOADING,
      })

      const config = {
        headers: {
          'Content-Type': 'application/json',
        },
      }

      const body = JSON.stringify({
        ...newTaskData,
      })

      let res = await axios.post('/api/task/', body, config)

      const newTask: TaskType = {
        id: res.data._id,
        title: res.data.title,
        project: res.data.project._id,
        time: res.data.time,
        done: res.data.done,
        date: new Date(res.data.date),
      }

      dispatch({
        type: CREATE_TASK_SUCCESS,
        payload: newTask,
      })
    } catch (error) {
      const errors = error.response.data.errors

      if (errors) {
        errors.forEach((error: any) => {
          dispatch(addAlert(AlertVariants.error, error.msg))
        })
      }

      dispatch({
        type: CREATE_TASK_FAIL,
      })
    }
  }

export const deleteTask =
  (id: string) => async (dispatch: Dispatch<TaskActionTypes>) => {
    try {
      dispatch({
        type: DELETE_TASK_LOADING,
      })

      const res = await axios.delete(`/api/task/${id}`)

      const task: TaskType = {
        id: res.data._id,
        title: res.data.title,
        project: res.data.project._id,
        time: res.data.time,
        done: res.data.done,
        date: new Date(res.data.date),
      }

      dispatch({
        type: DELETE_TASK_SUCCESS,
        payload: task.id,
      })
    } catch (e) {
      dispatch({
        type: DELETE_TASK_FAIL,
      })
    }
  }
