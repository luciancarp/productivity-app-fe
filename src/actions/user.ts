import { Dispatch } from 'redux'
import axios from 'axios'

import {
  GET_USER_SUCCESS,
  GET_USER_LOADING,
  GET_USER_FAIL,
  UserActionTypes,
  UserType,
  LOGIN_USER_LOADING,
  LOGIN_USER_SUCCESS,
  LOGIN_USER_FAIL,
  CREATE_USER_LOADING,
  CREATE_USER_SUCCESS,
  CREATE_USER_FAIL,
  AlertActionTypes,
  AlertVariants,
} from './types'

import setAuthToken from '../utils/setAuthToken'

import { addAlert } from './alert'

export const getUser = () => async (dispatch: Dispatch<UserActionTypes>) => {
  try {
    dispatch({
      type: GET_USER_LOADING,
    })

    // if (localStorage.token) {
    //   setAuthToken(localStorage.token)
    // }

    const res = await axios.get('/api/user')

    const user: UserType = {
      name: res.data.name,
      email: res.data.email,
      date: new Date(res.data.date),
    }

    dispatch({
      type: GET_USER_SUCCESS,
      payload: user,
    })
  } catch (error) {
    console.log(error)
    dispatch({
      type: GET_USER_FAIL,
    })
  }
}

export const loginUser =
  (loginData: { email: string; password: string }) =>
  async (dispatch: Dispatch<UserActionTypes>) => {
    try {
      dispatch({
        type: LOGIN_USER_LOADING,
      })

      const config = {
        headers: {
          'Content-Type': 'application/json',
        },
      }

      const body = JSON.stringify({
        email: loginData.email,
        password: loginData.password,
      })

      let res = await axios.post('/api/user/login', body, config)

      if (res.data.token) {
        localStorage.setItem('token', res.data.token)
        setAuthToken(res.data.token)
      }

      res = await axios.get('/api/user')

      const user: UserType = {
        name: res.data.name,
        email: res.data.email,
        date: new Date(res.data.date),
      }

      dispatch({
        type: LOGIN_USER_SUCCESS,
        payload: user,
      })
    } catch (error) {
      // const errors = error.response.data.errors

      // if (errors) {
      //   errors.forEach((error) => dispatch(setAlert(error.msg, 'error')))
      // }

      dispatch({
        type: LOGIN_USER_FAIL,
      })
    }
  }

export const createUser =
  (signUpData: { name: string; email: string; password: string }) =>
  async (dispatch: Dispatch<UserActionTypes> & Dispatch<AlertActionTypes>) => {
    try {
      dispatch({
        type: CREATE_USER_LOADING,
      })

      const config = {
        headers: {
          'Content-Type': 'application/json',
        },
      }

      const body = JSON.stringify({
        name: signUpData.name,
        email: signUpData.email,
        password: signUpData.password,
      })

      let res = await axios.post('/api/user/', body, config)

      if (res.data.token) {
        localStorage.setItem('token', res.data.token)
        setAuthToken(res.data.token)
      }

      res = await axios.get('/api/user')

      const user: UserType = {
        name: res.data.name,
        email: res.data.email,
        date: new Date(res.data.date),
      }

      dispatch({
        type: CREATE_USER_SUCCESS,
        payload: user,
      })
    } catch (error) {
      const errors = error.response.data.errors

      if (errors) {
        errors.forEach((error: any) => {
          dispatch(addAlert(AlertVariants.error, error.msg))
        })
      }

      dispatch({
        type: CREATE_USER_FAIL,
      })
    }
  }
