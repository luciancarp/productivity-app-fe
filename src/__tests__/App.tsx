import React from 'react'
import { render, screen, cleanup } from '@testing-library/react'
import App from '../App'
import tokenUtility from '../utils/tokenUtility'
import { logoutUser } from '../actions/user'

// test('renders learn react link', () => {
//   render(<App />)
//   const linkElement = screen.getByText(/learn react/i)
//   expect(linkElement).toBeInTheDocument()
// })

describe('App', () => {
  afterEach(cleanup)

  it('takes a snapshot', async () => {
    const { asFragment } = render(<App />)

    expect(asFragment()).toMatchSnapshot()
  })

  // TODO
  // it('sets auth token if it is stored in localStorage', () => {})
  // it('fetches user info', () => {})
  // it('contains a Provider component', () => {})
  // it('contains a ThemeProvider component', () => {})
  // it('contains a GlobalStyles component', () => {})
  // it('contains an Alerts component', () => {})
})
