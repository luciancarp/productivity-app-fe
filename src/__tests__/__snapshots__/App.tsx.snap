// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App takes a snapshot 1`] = `
<DocumentFragment>
  <div
    class="Toastify"
  />
  <div
    class="sc-jJoQpE jYLPoN"
  >
    <div
      class="sc-gWXaA-D iNHUpL"
    >
      <header
        class="sc-giYgFv jvFgBx"
      >
        <div
          class="sc-ikJzcn hgaiKi"
        >
          <h1
            class="sc-kLwgWK jskzYa"
          >
            Projects
          </h1>
          <button
            class="sc-eCImvq kUxYQR"
            type="button"
          >
            <h3
              class="sc-jRQAMF bXdgTm"
            >
              âž•
            </h3>
          </button>
        </div>
        <ul
          class="sc-ezbkgU bdKttU"
        />
      </header>
      <main
        class="sc-hiCivh jMaUs"
      />
      <aside
        class="sc-fKVsgm iBfyaE"
      >
        <div
          class="sc-iwjezw jVmKbk"
        >
          <a
            class="sc-dlVyqM dMIEZR"
            href="/login"
            width="100%"
          >
            <h3
              class="sc-kfPtPH kkMnqs"
            >
              Login
            </h3>
          </a>
        </div>
        <div
          class="sc-bBHwJV dCPqby"
        >
          <button
            class="sc-eCImvq kUxYQR"
            type="button"
          >
            <h3
              class="sc-jRQAMF bXdgTm"
            >
              ðŸŒ™
            </h3>
          </button>
        </div>
      </aside>
    </div>
  </div>
</DocumentFragment>
`;
